diff -ur ../gcc-14.1.0-clean/gcc/configure ./gcc/configure
--- ../gcc-14.1.0-clean/gcc/configure	2024-05-07 07:52:51.000000000 +0100
+++ ./gcc/configure	2024-07-17 15:22:37.385656470 +0100
@@ -5546,6 +5546,9 @@
     sh*-*-netbsd*)
 	PICFLAG_FOR_TARGET=-fpic
 	;;
+    m68k-*)
+        #nope
+        ;;
     # Default to -fPIC unless specified otherwise.
     *)
 	PICFLAG_FOR_TARGET=-fPIC
diff -ur ../gcc-14.1.0-clean/libgcc/configure ./libgcc/configure
--- ../gcc-14.1.0-clean/libgcc/configure	2024-05-07 07:51:41.000000000 +0100
+++ ./libgcc/configure	2024-07-17 15:21:17.980572171 +0100
@@ -2426,6 +2426,9 @@
     sh*-*-netbsd*)
 	PICFLAG=-fpic
 	;;
+    m68k-*)
+        #nope
+        ;;
     # Default to -fPIC unless specified otherwise.
     *)
 	PICFLAG=-fPIC
diff -ur ../gcc-14.1.0-clean/libiberty/configure ./libiberty/configure
--- ../gcc-14.1.0-clean/libiberty/configure	2024-05-07 07:51:41.000000000 +0100
+++ ./libiberty/configure	2024-07-17 15:22:49.865826820 +0100
@@ -5380,6 +5380,9 @@
     sh*-*-netbsd*)
 	PICFLAG=-fpic
 	;;
+    m68k-*)
+        #nope
+        ;;
     # Default to -fPIC unless specified otherwise.
     *)
 	PICFLAG=-fPIC
